BUFFER_CORE: Add clearFrame function.

lsouchet opened the pull request.

sbarthelemy commented on tf2/include/tf2/buffer_core.h:
```diff
@@ -103,6 +103,15 @@ class BufferCore
+   * This function will:
+   *  - call clearFrame() on each TimeCacheInterface.
+   *  - Delete the entries corresponding to the target frame in frameIDs_,
+   * frameIDs_reverse and frame_authority_
```
wrong indent on this line?

lsouchet replied:
done

sbarthelemy commented on tf2/include/tf2/time_cache.h:
```diff
@@ -106,14 +113,23 @@ class TimeCache : public TimeCacheInterface
+
+  /** This method will:
+   *  - remap the parent frame of a TransformStorage whose parent
+   * is the targeted frame to "NO_PARENT"
```
indent?

lsouchet replied:
not really.

lsouchet replied:
done

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -183,15 +184,55 @@ void BufferCore::clear()
+    for (M_TimeCacheInterface::iterator cache_it = frames_.begin();
+         cache_it != frames_.end(); )
+    {
+      if (cache_it->second)
```
&& the two if?

lsouchet replied:
done

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -183,15 +184,55 @@ void BufferCore::clear()
+  if (authority_it != frame_authority_.end())
+  {
+    frame_authority_.erase(authority_it);
+  }
```
I think

  frame_authority_.erase(frame_compact_id); 

would do it

lsouchet replied:
done

jlanca commented on tf2/include/tf2/buffer_core.h:
```diff
@@ -317,19 +326,24 @@ class BufferCore
+  /** \brief A frame counter used to increment the id of the newly created
+   * frames, even after deletion of some.
+  */
+  unsigned int frame_counter_;
```
Can you add a comment to explain that frame_counter_ != frameIDs_.size() because it keeps increasing even if you clear a frame?

jlanca commented on tf2/include/tf2/buffer_core.h:
```diff
@@ -317,19 +326,24 @@ class BufferCore
   M_StringToCompactFrameID frameIDs_;
   /** \brief A map from CompactFrameID frame_id_numbers to string for debugging and output */
-  std::vector<std::string> frameIDs_reverse;
+  typedef std::map<CompactFrameID, std::string> M_CompactFrameIDToString;
```
If you don't really need to iterate over the elements of the frameIDs_reverse map, maybe an unordered_map would be enough?

lsouchet replied:
you are right

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -282,14 +323,26 @@ bool BufferCore::setTransform(const geometry_msgs::TransformStamped& transform_i
+  }
+  if (it != frames_.end())
+  {
+    frames_.at(cfid) = newFrame;
```
*it = newFrame

lsouchet replied:
done

sebastiendalibard commented on tf2/src/buffer_core.cpp:
```diff
@@ -282,14 +323,26 @@ bool BufferCore::setTransform(const geometry_msgs::TransformStamped& transform_i
+  {
+    newFrame = TimeCacheInterfacePtr(new TimeCache(cache_time_));
+  }
+  if (it != frames_.end())
```
Why not just:

frames_[cfid] = newFrame;

lsouchet replied:
done

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -282,14 +323,26 @@ bool BufferCore::setTransform(const geometry_msgs::TransformStamped& transform_i
   }
 
-  return frames_[cfid];
+  return frames_.at(cfid);
```
please use the ngaga style:

```
  std::pair<iterator,bool> ret = frames_.insert(std::make_pair(cfid, newFrame));
  if (!ret.second)
    *ret.first =  newFrame;
  return *ret.first;
```

lsouchet replied:
did not know about this convention :P

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -843,14 +897,15 @@ CompactFrameID BufferCore::lookupOrInsertFrameNumber(const std::string& frameid_
     }
     else
-      return frameIDs_reverse[frame_id_num];
+      return frameIDs_reverse.at(frame_id_num);
```
return *frameIt

lsouchet replied:
done

sbarthelemy commented on tf2/include/tf2/buffer_core.h:
```diff
@@ -317,19 +326,24 @@ class BufferCore
-  typedef std::vector<TimeCacheInterfacePtr> V_TimeCacheInterface;
-  V_TimeCacheInterface frames_;
-  
+  typedef std::map<CompactFrameID, TimeCacheInterfacePtr> M_TimeCacheInterface;
```
flat_map!

lsouchet replied:
done

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -803,11 +856,12 @@ bool BufferCore::canTransform(const std::string& target_frame, const ros::Time&
   else
   {
-    return frames_[frame_id];
+    return it->second;
```
yessss!

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -803,11 +856,12 @@ bool BufferCore::canTransform(const std::string& target_frame, const ros::Time&
+  if (it == frames_.end())
     return TimeCacheInterfacePtr();
   else
   {
```
rm "{}"?

lsouchet replied:
I don't really like but whatever

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -873,25 +928,32 @@ std::string BufferCore::allFramesAsString() const
-    mstream << "Frame "<< frameIDs_reverse[counter] << " exists with parent " << frameIDs_reverse[frame_id_num] << "." <<std::endl;
+    try {
+      mstream << "Frame "<< child_id_num << " (" << frameIDs_reverse.at(child_id_num)<< ") exists with parent " <<
+                 frame_id_num << " (" << frameIDs_reverse.at(frame_id_num) << "." << ") at " <<
```
"." is at the wrong place

lsouchet replied:
done

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -183,15 +184,55 @@ void BufferCore::clear()
+    {
+      if (cache_it->second)
+      {
+        if (cache_it->second->clearFrame(frame_compact_id))
```
better erase  and don't return a bool

lsouchet replied:
done

sbarthelemy commented on tf2/include/tf2/buffer_core.h:
```diff
@@ -317,19 +328,24 @@ class BufferCore
 
   /** \brief A map from string frame ids to CompactFrameID */
-  typedef boost::unordered_map<std::string, CompactFrameID> M_StringToCompactFrameID;
+  typedef boost::container::flat_map<std::string, CompactFrameID> M_StringToCompactFrameID;
```
I see no reason for this container change.

lsouchet replied:
done

sbarthelemy commented on tf2/include/tf2/buffer_core.h:
```diff
@@ -317,19 +328,24 @@ class BufferCore
   M_StringToCompactFrameID frameIDs_;
   /** \brief A map from CompactFrameID frame_id_numbers to string for debugging and output */
-  std::vector<std::string> frameIDs_reverse;
+  typedef boost::container::flat_map<CompactFrameID, std::string> M_CompactFrameIDToString;
```
this container change is nice!

lsouchet replied:
cool!

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -183,15 +184,50 @@ void BufferCore::clear()
+
+}
+
+void BufferCore::clearFrame(const std::string& frame_id_str)
```
what if frame_id_str == "NO_PARENT"?

lsouchet replied:
done

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -873,25 +921,32 @@ std::string BufferCore::allFramesAsString() const
     }
-    mstream << "Frame "<< frameIDs_reverse[counter] << " exists with parent " << frameIDs_reverse[frame_id_num] << "." <<std::endl;
+    try {
+      mstream << "Frame "<< child_id_num << " (" << frameIDs_reverse.at(child_id_num)<< ") exists with parent " <<
```
possibly use while not initialized. Is it different from compactFID?

lsouchet replied:
done

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -873,25 +921,32 @@ std::string BufferCore::allFramesAsString() const
     if(  frame_ptr->getData(ros::Time(), temp))
+    {
       frame_id_num = temp.frame_id_;
+      child_id_num = temp.child_frame_id_;
```
Is it different from compactFID?

lsouchet replied:
it should not be

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -1311,9 +1366,10 @@ void BufferCore::_getFrameStrings(std::vector<std::string> & vec) const
   //  for (std::vector< TimeCache*>::iterator  it = frames_.begin(); it != frames_.end(); ++it)
-  for (unsigned int counter = 1; counter < frameIDs_reverse.size(); counter ++)
+  for (M_CompactFrameIDToString::const_iterator it = frameIDs_reverse.begin();
+       it != frameIDs_reverse.end(); ++it)
```
I think you now include NO_PARENT.

lsouchet replied:
done

sbarthelemy commented on tf2/include/tf2/buffer_core.h:
```diff
@@ -103,6 +104,16 @@ class BufferCore
+  /** \brief Clear a specified frame
+   * \param frame_id_str The string name of the frame that will be deleted.
+   * This function will:
+   *  - clear the frames whose child is the frame to delete,
```
frameS or frame?

lsouchet replied:
done

sbarthelemy commented on tf2/include/tf2/buffer_core.h:
```diff
@@ -103,6 +104,16 @@ class BufferCore
   /** \brief Clear all data */
   void clear();
 
+  /** \brief Clear a specified frame
```
a general remark for this function:

* what about "eraseFrame()" instead of "clear()"? It is IMO more inline with std::map's own clear vs. erase functions.
* the doc should explain what it does but, not how it does it.

lsouchet replied:
There is already a clear() function, so I think it makes sense to keep a clearFrame() to express that they both serve a similar purpose.

lsouchet replied:
done

sbarthelemy commented on tf2/include/tf2/buffer_core.h:
```diff
@@ -103,6 +104,16 @@ class BufferCore
+   * \param frame_id_str The string name of the frame that will be deleted.
+   * This function will:
+   *  - clear the frames whose child is the frame to delete,
+   *  - call clearParentFrame() on each TimeCacheInterface,
```
Reparent each frame whose parent is frame_id_str to NO_PARENT.

lsouchet replied:
done

sbarthelemy commented on tf2/src/cache.cpp:
```diff
@@ -294,6 +294,19 @@ ros::Time TimeCache::getOldestTimestamp()
   return storage_.back().stamp_;
 }
 
+void tf2::TimeCache::clearParentFrame(tf2::CompactFrameID &id)
```
better name? unparentFrame?

lsouchet replied:
done

sbarthelemy commented on tf2/include/tf2/buffer_core.h:
```diff
@@ -103,6 +104,14 @@ class BufferCore
   void clear();
 
+  /** \brief Clear a specified frame
+   * \param frame_id_str The string name of the frame that will be deleted.
```
s/string//

lsouchet replied:
done

sbarthelemy commented on tf2/include/tf2/buffer_core.h:
```diff
@@ -103,6 +104,14 @@ class BufferCore
 
+  /** \brief Clear a specified frame
+   * \param frame_id_str The string name of the frame that will be deleted.
+   * This function will clear the frame whose child is the frame to delete and
```
what about:

This function will remove the frame, and reparent all its children frames to NO_PARENT.

lsouchet replied:
done

sbarthelemy commented on tf2/include/tf2/buffer_core.h:
```diff
@@ -103,6 +104,14 @@ class BufferCore
   /** \brief Clear all data */
   void clear();
 
+  /** \brief Clear a specified frame
```
s/clear/remove/ (or erase)

lsouchet replied:
done

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -183,15 +184,55 @@ void BufferCore::clear()
+    std::cerr << "Unable to erase NO_PARENT frame" << std::endl;
+    return;
+  }
+  if (!frames_.empty())
```
unneeded

lsouchet replied:
done

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -183,15 +184,55 @@ void BufferCore::clear()
+  frameIDs_reverse.erase(frame_compact_id);
+  frame_authority_.erase(frame_compact_id);
+  frameIDs_.erase(it);
+  testTransformableRequests();
```
what does it do?

lsouchet replied:
done

sbarthelemy commented on tf2/src/buffer_core.cpp:
```diff
@@ -1296,11 +1354,16 @@ void BufferCore::_getFrameStrings(std::vector<std::string> & vec) const
+  if (frameIDs_reverse.size() <= 1u)
+  {
+    return;
+  }
```
couldn't it be

  assert(!frameIDs_reverse.empty());

lsouchet replied:
not really because if the size is 1 (only NO_PARENT in the tree, I don't want to consider it.

sbarthelemy replied:
The "skip NO_PARENT" part is already dealt with by the begin()+1 below. You just need the assert to document that NO_PARENT won't go away.

lsouchet replied:
done

Commit:
BUFFER_CORE: Add erase function.

This includes:
- structure containers change to boost::flat_map to dynamically handle
  data access after a deletion.
- erase() function adding to clear the data from a frame entry

sbarthelemy reviewed (approved):
lgtm!
